### PR TITLE
Core/Movement: Prevent pet animation stuttering during movement

### PR DESCRIPTION
**Pull request by Stoabrogga**
##### CHANGES PROPOSED:
This is more of a base for discussion, although it works on my solo test-server.

It bugged me since starting with AC that the pets and vanity pets start to stutter while running beneath the character. The cause of this is a fixed speed calculation for the pet speed while not in combat, see line 13097 in Unit.cpp:

```
if (speed < pOwner->GetSpeedRate(mtype)+0.1f)
    speed = pOwner->GetSpeedRate(mtype)+0.1f; // pets derive speed from owner when not in combat
```

This leads to the pet constantly trying to outrun the player but has to wait if too far ahead. This wait is causing the animation to stutter.

As a possible fix I adjusted the "TargetedMovementGenerator" in order to update the speed of pets every tick and changed the speed calculation in Unit.cpp (only for pets and vanity pets). This works well for me, but I'm too inexperienced with AC to tell **if this could cause a greater performance issue.**

I also altered the distance and angle of vanity pets so that they are nearer and slightly behind the player.

##### ISSUES ADDRESSED:
none

##### TESTS PERFORMED:
Tested build on Ubuntu 16.04 and in-game

##### HOW TO TEST THE CHANGES:
Just summon pets or vanity pets and run or fly around.

##### KNOWN ISSUES AND TODO LIST:
none

##### Target branch(es):
Master

### CODE REVIEW

**Comment by BarbzYHOOL** [comment]
Very nice idea for your edit haha

**Comment by wetbrownsauce** [comment]
I'm still getting stuttering with my hunter pet. Vanity pets seem to move smoothly, but thats about it

**Comment by Stoabrogga** [comment]
I altered the speed algorithm for the normal pets a little bit. Could you please test it once more?

**Comment by wetbrownsauce** [comment]
So I still get stutter if I strafe with a pet following me, but there is no stutter when running straight forward. 

If I fly a couple feet off the ground, pet does not stutter, but begins stuttering if I fly any higher than that.

**Comment by BarbzYHOOL** [comment]
@Rochet2 what do you think of this in terms of performance with hundreds of players?

**Comment by Stoabrogga** [comment]
Ok, I changed the distance calculation once more and now differentiate between the player flying or running. But it would be great to know if performance could be a show-stopper here. At last the speed of the pets have to be updated regularly although I tried to simplify the algorithm and increase the range where no changes to the speed occur.

**Comment by wetbrownsauce** [comment]
I'm not seeing any stuttering anymore now. I can't comment on how it might affect performance, but other than that companions and pets seem to only stutter on rare occasion

**Comment by FrancescoBorzi** [comment]
is this ready to be merged?

**Comment by Stoabrogga** [comment]
@FrancescoBorzi 
The changes seem to work, but before merging someone who has more insight on the code than I do should check them in order to ensure that there won't be any performance issues.

**Comment by FrancescoBorzi** [comment]
@Viste @Nefertumm @Winfidonarleyan  can you please have a look at this ?

**Comment by Nefertumm** [comment]
I didn't see any potential issue. I think it's safe to merge 👍

**Comment by Viste** [comment]
Looks good i think we don't get any issues with this

**Comment by FrancescoBorzi** [comment]
thanks guys for your feedback

**Comment by BarbzYHOOL** [comment]
@Stoabrogga 

> it might slightly affect performance, but a priori I think it wouldn't be noticeable
> what bothers me though is if the changes were tested with all the classes in-game
> and pet + vanity pet
> how would it affect an enhancement shaman with both wolves out?

**Comment by Stoabrogga** [comment]
@BarbzYHOOL 
The change only affects out-of-combat distance and speed of the normal pets, not positioning (this is only changed for vanity pets). I did not test it with every class, but with warlock, death knight, hunter and a few other classes using the Beastmaster module, with and without vanity pet. I also did test the feral spirit of the shaman, and both wolves run on the left and right side of the player.

**Comment by BarbzYHOOL** [comment]
@Stoabrogga thank you for the PR :)

thank you all for input

File: src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp [hunk 1]
```diff
@@ -98,8 +98,17 @@ void TargetedMovementGeneratorMedium<T,D>::_setTargetLocation(T* owner, bool ini
         if ((!initial || (owner->movespline->Finalized() && this->GetMovementGeneratorType() == CHASE_MOTION_TYPE)) && i_target->IsWithinDistInMap(owner, dist) && i_target->IsWithinLOS(owner->GetPositionX(), owner->GetPositionY(), owner->GetPositionZ()))
             return;
 
-        // Xinef: Fix follow angle for hostile units
         float angle = i_angle;
+
+        if (i_target->GetTypeId() == TYPEID_PLAYER && owner->ToCreature()->GetCreatureType() == CREATURE_TYPE_NON_COMBAT_PET)
```
Comment (Rochet2) [inline src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp]: Is it somewhere checked that `owner->ToCreature()` is safe? Is the type checked somewhere?  I cannot see it.

Comment (FrancescoBorzi) [reply src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp]: @Stoabrogga

Comment (Stoabrogga) [reply src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp]: Yes, you're right, it is better to check if the owner is a creature before determining the creature type:
```
if (i_target->GetTypeId() == TYPEID_PLAYER)
{
	Creature* creature = owner->ToCreature();

	if (creature && creature->GetCreatureType() == CREATURE_TYPE_NON_COMBAT_PET)
	{
		// fix distance and angle for vanity pets
		dist = 0.3f;
		angle = PET_FOLLOW_ANGLE + M_PI * 0.2f;
		size = i_target->GetCombatReach() - i_target->GetObjectSize();
	}
}
```
Can someone commit this directly or should I create another PR for this?

Comment (FrancescoBorzi) [reply src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp]: please open a PR, we should avoid committing directly

**Pull request merged**